### PR TITLE
feat!: use native resolver in loader

### LINKED ISSUE #4672: js loader should use rust side resolver other than enhanced-resolve
currently js loader use enhanced-resolver which cause resolve.tsconfig not working for js loader like sass-loader, so we don't support sass alias now， see https://github.com/hardfist/rspack-webpack-nest-alias/pull/1

### PR DESCRIPTION
**Pull request by bvanjoi**
Fixes #4672

### CODE REVIEW

File: crates/rspack_binding_values/src/compilation.rs [hunk 1]
```diff
@@ -428,6 +430,21 @@ impl JsCompilation {
...
+  #[napi]
+  pub fn create_resolver(&mut self, options: raw_resolve::RawResolveOptions) -> JsResolver {
+    let options = rspack_core::ResolveOptionsWithDependencyType {
+      resolve_options: Some(Box::new(
+        options
+          .try_into()
+          .expect("the options of resolver is invalid"),
+      )),
+      resolve_to_context: false,
+      dependency_category: rspack_core::DependencyCategory::CommonJS,
```
Comment (ahabhgk) [inline crates/rspack_binding_values/src/compilation.rs:443]: are you sure?

File: packages/rspack/src/NativeResolverFactory.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			callback(null, false);
+		} else if (res.startsWith(ERROR_PREFIX)) {
+			const error = Error(res.slice(ERROR_PREFIX.length));
+			callback(error);
+		} else {
+			callback(null, res);
+		}
+	}
+	doResolve() {
+		throw Error("The method fo `doResolve` is not available in nativeResolver");
```
Comment (ahabhgk) [inline packages/rspack/src/NativeResolverFactory.ts:88]: fo -> of

File: packages/rspack/src/Compilation.ts [hunk 3]
```diff
@@ -130,6 +132,8 @@ export class Compilation {
 	outputOptions: OutputNormalized;
 	compiler: Compiler;
 	resolverFactory: ResolverFactory;
+	// TODO: maybe we can replace `resolverFactory` to `nativeResolverFactory`
+	nativeResolverFactory: NativeResolverFactory;
```
Comment (ahabhgk) [inline packages/rspack/src/Compilation.ts:136]: Do we still need to keep resolverFactory?

Comment (bvanjoi) [reply packages/rspack/src/Compilation.ts:136]: I'm not sure, but I intend to keep it.

**Comment by hardfist** [comment]
@bvanjoi I don't remember why we don't hold on this?

File: packages/rspack/src/NativeResolverFactory.ts [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	get hooks(): any {
+		throw Error("The filed of `hooks` is not available in nativeResolver");
+	}
+	ensureHook(): any {
+		throw Error(
+			"The method of `ensureHook` is not available in nativeResolver"
+		);
+	}
+	getHook(): any {
+		throw Error("The method of `getHook` is not available in nativeResolver");
```
Comment (bvanjoi) [inline packages/rspack/src/NativeResolverFactory.ts:55]: @hardfist There are many incompatible api.

**Comment by hardfist** [comment]
close for outdated

**Comment by SyMind** [comment]
!bench

**Review by JSerFeng — APPROVED** [review]

**Comment by hardfist** [comment]
this is also breaking change, please note add breaking change notice

**Comment by chenjiahan** [comment]
Hi, I wonder in which case this will break the user's config or plugin, can we add more explanations

**Comment by chenjiahan** [comment]
Added "release: breaking change" label for changelog generating

**Comment by ahabhgk** [comment]
!eco-ci

**Comment by chenjiahan** [comment]
It looks like an Rsbuild e2e case is failed:

image

**Pull request merged**